### PR TITLE
WIP WIP: Add the ability to blacklist a service instance

### PR DESCRIPTION
**Pull request by Raynos**
This is a WIP PR to add the foundation for blacklisting
a service instance.
- Update the thrift file
- Create a new hyperbahn-service.js that implements blacklist;
  eventually we'll move all of handler.js into it and
  clean it up.
- Create a BlacklistInstance that manages the state and the
  timing aspects.
- Leave TODOs about HyperbahnServices interacts with
  EgressNodes & ServiceProxy.

TODO: 
- [ ] tests
- [ ] remove & drain peer in ServiceProxy
- [ ] fanout on EgressNodes change
- [ ] validation & other features

r: @jcorbin @rf

### CODE REVIEW

File: hyperbahn-service.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    // TODO: listen to egressNodes and fanout to re-blacklist ?
+}
+
+HyperbahnService.prototype.blacklist =
+function blacklist(req, headers, body, cb) {
+    var self = this;
+
+    // TODO: validate body.serviceName
+    // TODO: validate body.instanceHostPort
+    // TODO: validate body.timeToBan
```
Comment (jcorbin) [inline hyperbahn-service.js]: it's thrift, how are they not already validated

**Comment by jcorbin** [comment]
Please "ban" something, e.g. "banPeer", "banWorker", etc

**Comment by jcorbin** [comment]
to me "blacklist" is indefinite without expiry, where-as "ban" seems much more sensible to be time-bounded.

File: hyperbahn-service.js (hunk 1 above)
Comment (Raynos) [inline hyperbahn-service.js]: edge cases:

 - empty string
 - non-valid string format host:port
 - negative number / zero.

File: handler.js [hunk 2]
```diff
@@ -127,11 +142,28 @@ HyperbahnHandler.prototype.type = 'hyperbahn.advertisement-handler';
 */
 HyperbahnHandler.prototype.handleAdvertise =
 function handleAdvertise(handler, req, arg2, arg3, cb) {
-    handler.sendRelays(req, arg2, arg3, 'relay-ad', cb);
+    var services = arg3.services;
+    var fanoutServices = [];
+
+    for (var i = 0; i < services.length; i++) {
```
Comment (rf) [inline handler.js]: Could we do this logic in `#sendRelays` instead? There's already a loop over the host ports there that adds them to a new data structure, would be nice to only do that once.

**Comment by Raynos** [comment]
RIP

**Pull request closed without merging**